File: JavaScript/프로그래머스/2/43165. 타겟 넘버/타겟 넘버.js
// // dfs
// function solution(numbers, target) {
//     let count =0;
    
//     function dfs (index, sum) {
//         if(index === numbers.length) {
//             if(sum === target) count++
//             return;
//         }
//         dfs(index + 1, sum + numbers[index]);
//         dfs(index + 1, sum - numbers[index]);
//     }
//     dfs(0,0)
//     return count;
// }


// bfs
function solution(numbers, target) {
    let count = 0;
    let queue = [[0, 0]]; // [현재 합계, 인덱스]

    for (let i = 0; i < numbers.length; i++) {
        let newQueue = [];

        for (let [sum, index] of queue) {
            newQueue.push([sum + numbers[index], index + 1]);
            newQueue.push([sum - numbers[index], index + 1]);
        }

        queue = newQueue; // 큐를 갱신하여 크기를 제한
    }

    return queue.filter(([sum, index]) => index === numbers.length && sum === target).length;
}

File: Java/프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.LinkedList;
import java.util.Queue;


class Solution {
    public int solution(int[] numbers, int target) {
        int count = 0;
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{0,0});
        
        while(!queue.isEmpty()) {
            int[] current = queue.poll();
            int sum = current[0];
            int index = current[1];
            
            if(index == numbers.length) {
                if(sum == target) count++;
                continue;
            }
            queue.add(new int[] {sum + numbers[index], index+1});
            queue.add(new int[] {sum - numbers[index], index+1});
        }        
        return count;
    }
}